### DISCUSSION TITLE
Is 2FA supported?

### DISCUSSION THREAD

**Discussion by Alfetta159** [jdhitsolutions/PSBluesky#21] [Q&A]
I get a message saying that a auth code has been sent to my email, but I'm errored out before I can do anything about it.

```pwsh
PS /home/danp> Get-BskyAccessToken

cmdlet Get-BskyAccessToken at command pipeline position 1
Supply values for the following parameters:
(Type !? for Help.)
Credential
User: alfetta159.bsky.social 
Password for user alfetta159.bsky.social: *********

Invoke-RestMethod: /home/danp/.local/share/powershell/Modules/PSBlueSky/1.3.0/functions/helpers.ps1:171
Line |
 171 |          $r = Invoke-RestMethod @splat
     |               ~~~~~~~~~~~~~~~~~~~~~~~~
     |  {   "error": "AuthFactorTokenRequired",   "message": "A sign in code
     | has been sent to your email address" }
PS /home/danp> $PSDefaultParameterValues['Alfetta159-Bsky.Social:Credential'] = $BlueskyCredential
PS /home/danp> 
```

**Comment by jdhitsolutions** [comment]
I just updated the module to v2..0.0 and revised this process. Now you would run `Start-BskySession` once before running most any module command. The module does not support 2FA. That is something I will have to investigate.

**Comment by jdhitsolutions** [comment]
I'd like to get feedback from others on this topic. I found this [background information](https://github.com/bluesky-social/atproto/discussions/2435). I think to support email auth, the new session process would need to wait for the user to enter the authentication token. I wonder if moving to an AppPassword is a better long term solution. That would eliminate the need for credentials.

**Comment by OS-Scratch** [comment]
2FA is supported, but not for e-mail. Use mobile apps, browser extensions, push notifications or passkeys instead.

**Reply by jdhitsolutions** [reply]
Right now, you can configure your Bluesky account to use 2FA via email. When you logon, you'll get a verification code in your email which you then paste into the app. When using the API, you would have to resend the body with the verification code. For this module, the way I envision it is to create the session and capture any exception. If the exception indicates the need for an authentication code, the function would use `Read-Host` to accept the emailed code. The function would update the body and resubmit the API request. This isn't unreasonable,  but if the current 2FA setup isn't used by many people, I'd rather not spend the time.

The better approach, and one that I will be pushing in the next release, is to use an app password. You can still use 2FA, but an app password doesn't require it and you can cancel the app password at anytime.

**Reply by jdhitsolutions** [reply]
Currently, this module doesn't support 2FA. I have an idea of how it might be implemented but I haven't started down that path. I am thinking that an app password is the best approach here as this would never expose your real password.

**Reply by Alfetta159** [reply]
If it helps anyone else, I wasn't aware until very recently that Bluesky had an app password generator. It hadn't occurred to me to try that.

[https://bsky.app/settings/app-passwords](https://bsky.app/settings/app-passwords)

**Comment by jhoneill** [comment]
With the current design I suspect 2FA will be very clunky, because the refresh token is not saved, so every time a new session starts the module goes back to user name and plain text password in an https request body.  Every time that happens you get the extra authentication step before the module can get an access token (and refresh token); then every 15 minutes this one uses referesh token to get a new access token and it's OK for the current session, but the next time it's back to name, password and extra factor  

I have taken to saving the refresh token (secure string in clixml, so it can only be read by my account on my computer, so access to my laptop is the authentication factor, which gives access to my mail and potentially opens everything.) 
The refresh token and access token is read at that start of the session if the AT is too old we get a new one and saves the new tokens for next time. (There are complications when I have multiple processes both thinking they own the same refresh token, but I've worked round them)

I think something similar is needed to make 2FA usable in a module, otherwise you have an interactive step each time its used which narrows the usefulness a lot.

**Reply by jdhitsolutions** [reply]
I think my refresh method would work with 2FA but I would rather have people use an app password. I don’t see any reason to support 2FA support in this module.

**Comment by jhoneill** [comment]
I agree.  Refresh should be fine regardless of the auth method, but the token doesn't carry over when powershell exits, so the next session is a reauth, 2FA cuts in again, and refresh until the end of the session
